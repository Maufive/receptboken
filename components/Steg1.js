import React, { Component } from "react";
import PropTypes from "prop-types";
import { Button, DisabledButton } from "./styles/Button";
import {
	RecipeForm,
	ListItemDiv,
	Header,
	List,
	IngrediensWrapper
} from "./styles/Steg1Styles";

class Steg1Test extends Component {
	state = {
		arr: [], // en array med ingredienser
		item: {
			input: "",
			numberOfUnits: 1,
			units: "st"
		},
		servings: 2,
		step: 1,
		editing: false,
		editIndex: null
	};

	componentDidMount() {
		this.setState({
			arr: this.props.ingredients,
			servings: this.props.servings
		});
	}

	addToArray = async e => {
		// Stoppa knappen fr친n att skicka en request
		e.preventDefault();
		// Ta en kopia av objektet Item i state
		const item = { ...this.state.item };
		// Kolla s친 att det inte redan finns en likadan ingrediens sparad redan f칬r att f칬rhindra dubletter
		const doubles = this.state.arr.filter(ingr => ingr.input === item.input);
		if (doubles.length >= 1)
			return this.props.setMessage("info", "Du har redan en s친dan i listan 游뱈");
		if (item.input.length >= 2) {
			// efter state blivit sparat, t칬m inputf칛ltet
			const emptyInputField = { ...this.state.item };
			emptyInputField.input = "";
			emptyInputField.numberOfUnits = 1;
			// ta en kopia av arrayen i state, och l칛gg till item objektet i den nya arrayen innan den sparas
			// samt t칬mmer inputen
			await this.setState(prevState => ({
				arr: [...prevState.arr, item],
				item: emptyInputField
			}));
		} else {
			this.props.setMessage("info", "Ingrediensen m친ste ha minst 2 tecken!");
			return;
		}
	};

	deleteItem = e => {
		// 1. G칬r en kopia av arrayen i state
		let arr = [...this.state.arr];
		// 2. Hitta r칛tt item att ta bort
		let index = arr.indexOf(e);
		// 3. Ta bort itemet fr친n arrayen
		if (index !== -1) {
			arr.splice(index, 1);
			this.setState({ arr });
		}
	};

	// leta fram objektet som ska editas, och s칛tt det i state
	editItem = currentItem => {
		// 1. G칬r en kopia av arrayen med ingredienser i state
		let items = [...this.state.arr];
		// 2. G칬r en kopia av det item jag vill 칛ndra
		let itemIndex = items.indexOf(currentItem);
		// 3. Ta en kopia av objektet som 칛r p친 den indexen
		const item = items[itemIndex];
		// 4. S칛tt state med det itemet som ska editas
		this.setState({
			item,
			editing: true,
			editIndex: itemIndex
		});
	};

	saveEdit = e => {
		const { item } = this.state;
		const index = this.state.editIndex;
		e.preventDefault();
		// kolla s친 det finns n친got i inputen
		if (item.input.length >= 2) {
			// ta en kopia p친 arrayen i state f칬r att modifiera
			const arr = [...this.state.arr];
			// spara 칬ver objektet i arrayen med objektet i state. Beh칬ver index.
			arr[index] = item;
			// cleara ut inputen
			const emptyInputField = { ...item };
			emptyInputField.input = "";
			emptyInputField.numberOfUnits = 1;
			// spara.
			this.setState({
				arr,
				item: emptyInputField,
				editing: false
			});
		} else {
			this.hideShowMessage("Ingrediensen m친ste ha minst 2 tecken!");
		}
	};

	saveIngredient = e => {
		const item = { ...this.state.item };
		item.input = e.target.value;
		this.setState({ item });
	};

	handleUnits = e => {
		let item = { ...this.state.item };
		item.units = e.target.value;
		this.setState({ item });
	};

	saveNumberOfUnits = e => {
		const item = { ...this.state.item };
		item.numberOfUnits = e.target.value;
		this.setState({ item });
	};

	handleServings = e => {
		this.setState({ servings: e.target.value });
	};

	// S칛tter state i parent-component med v칛rdet f칬r alla ingredienser. Detta kommer g칬ra att denna component unmountar och en
	// component f칬r n칛sta steg kommer att mountas ist칛llet.
	nextStep = e => {
		e.preventDefault();
		this.props.saveIngredients(this.state.arr, this.state.servings);
	};

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	render() {
		const { step } = this.state;
		return (
			<RecipeForm>
				<Header>
					<input
						type="text"
						value={this.props.title}
						onChange={this.props.saveToState}
						name="title"
						required
						placeholder="Nytt recept..."
					/>
					<i className="icofont-ui-edit" />
				</Header>
				<IngrediensWrapper>
					<h3>{step}. Ingedienser</h3>
					<div>
						<label htmlFor="servings">Antal portioner: </label>
						<select value={this.state.servings} onChange={this.handleServings}>
							<option value="2">2</option>
							<option value="4">4</option>
							<option value="6">6</option>
							<option value="8">8</option>
						</select>
					</div>
					<label htmlFor="item">Fyll i ingredienser:</label>
					<input
						type="text"
						name="item"
						value={this.state.item.input}
						onChange={this.saveIngredient}
						placeholder="Spaghetti..."
					/>
					<label htmlFor="numberOfUnits">Antal: </label>
					<input
						type="number"
						name="numberOfUnits"
						value={this.state.item.numberOfUnits}
						onChange={this.saveNumberOfUnits}
						style={{
							width: 60
						}}
					/>
					<label htmlFor="units">Enhet: </label>
					<select value={this.state.item.units} onChange={this.handleUnits}>
						<option defaultValue="st">Styck</option>
						<option value="gram">Gram</option>
						<option value="msk">Matsked</option>
						<option value="tsk">Tesked</option>
						<option value="dl">Deciliter</option>
					</select>
					{/* Om anv칛ndaren redigerar ett item, 칛ndra knappen till 츿ndra ist칛llet f칬r L칛gg till */}
					<div>
						{this.state.editing ? (
							<Button onClick={e => this.saveEdit(e, this.state.editIndex)}>
								<i className="icofont-ui-edit" />
								츿ndra
							</Button>
						) : (
							<Button onClick={this.addToArray}>
								<i className="icofont-plus" />
								L칛gg till
							</Button>
						)}
					</div>
				</IngrediensWrapper>
				<List>
					{this.state.arr.map(item => (
						<li key={item.input}>
							<ListItemDiv>
								<p>
									游녤 <span>{item.numberOfUnits}</span> <span>{item.units}</span>{" "}
									{item.input}
								</p>
								<div>
									<i
										onClick={() => this.deleteItem(item)}
										className="icofont-trash"
									/>
									<i
										onClick={() => this.editItem(item)}
										className="icofont-edit"
									/>
								</div>
							</ListItemDiv>
						</li>
					))}
				</List>
				{this.state.arr.length <= 1 && (
					<DisabledButton>
						{/* Beh칬ver en hover-effekt f칬r att indikera att man g친r vidare till n칛sta steg */}
						<i className="icofont-ui-next" />
						N칛sta steg
					</DisabledButton>
				)}
				{this.state.arr.length >= 2 && (
					<Button primary onClick={this.nextStep}>
						{/* Beh칬ver en hover-effekt f칬r att indikera att man g친r vidare till n칛sta steg */}
						<i className="icofont-ui-next" />
						N칛sta steg
					</Button>
				)}
			</RecipeForm>
		);
	}
}

export default Steg1Test;

Steg1Test.propTypes = {
	saveToState: PropTypes.func,
	toNextStep: PropTypes.func,
	ingredients: PropTypes.array,
	servings: PropTypes.number
};
