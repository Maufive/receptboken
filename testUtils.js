export const fakeRecept = {
	_id: "1234",
	author: "5678436374re67fwuf76",
	created: "Idag",
	description: ["Steg1", "Steg2", "Steg3"],
	id: "1337",
	ingredients: [
		{
			input: "Lök",
			numberOfUnits: "2",
			units: "dl"
		}
	],
	photo: "asdf123.jpg",
	reviews: [{ score: 1 }],
	servings: "2",
	tags: ["Mat", "Gott"],
	timeRequired: "20 min",
	title: "Mitt jättegoda recept"
};

export const fakeUser = {
	_id: "1234",
	username: "fakeuser@gmail.com",
	password: "1234",
	hearts: [],
	fname: "Niklas",
	lname: "Albinsson",
	description: "Hej jag heter Niklas",
	photo: "http://fakephoto.jpg"
};

export const fakeShoppinglist = [
	{
		_id: "34567876543452176",
		title: "My Shoppinglist",
		created: "12/3 2019",
		author: "61386712376576d7q6ewd65",
		list: [
			{
				input: "Lök",
				numberOfUnits: "5",
				units: "dl"
			},
			{
				input: "Tomater",
				numberOfUnits: "10",
				units: "st"
			}
		]
	}
];
