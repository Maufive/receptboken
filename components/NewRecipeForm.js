import React, { Component } from "react";
import { Header, RecipeForm } from "./styles/Steg1Styles";
import Steg1 from "./Steg1";
import Steg2 from "./Steg2";
import Steg3 from "./Steg3";

class NewRecipeForm extends Component {
	state = {
		title: "",
		description: [],
		tags: [],
		timeRequired: 0,
		photo: "",
		ingredients: [],
		servings: 2,
		step: 1
	};

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	previousStep = () => {
		this.setState({ step: this.state.step - 1 });
	};

	saveIngredients = (arr, servings) => {
		this.setState({
			ingredients: arr,
			servings,
			step: this.state.step + 1
		});
	};

	saveDescription = arr => {
		this.setState({
			description: arr,
			step: this.state.step + 1
		});
	};

	saveDetails = (photo, time, tags) => {
		this.setState({
			photo,
			timeRequired: time,
			tags,
			step: this.state.step + 1
		});
	};

	render() {
		const step = this.state.step;
		let ActiveComponent;
		// För varje steg i processen att skapa ett nytt recept så renderar jag en specifik component.
		// Syftet är att dela upp en lång process att ladda upp ett recept i mindre bitar för att göra det
		// lättare för användare.
		if (step === 1) {
			ActiveComponent = (
				<Steg1
					saveToState={this.saveToState}
					saveIngredients={this.saveIngredients}
					ingredients={this.state.ingredients}
					servings={this.state.servings}
				/>
			);
		} else if (step === 2) {
			ActiveComponent = (
				<Steg2
					saveDescription={this.saveDescription}
					steg={step}
					previousStep={this.previousStep}
					description={this.state.description}
				/>
			);
		} else if (step === 3) {
			ActiveComponent = (
				<Steg3
					steg={step}
					previousStep={this.previousStep}
					saveDetails={this.saveDetails}
				/>
			);
		} else {
			ActiveComponent = <h2>Finished!</h2>;
			// To-do: Skriv en funktion för en POST-method att passa ner till steg 3. Spara Allt som finns i state här
			// till databasen.
			// Gör en LOADING komponent som snurrar medans man väntar på att databasen ska spara receptet
			// Gör en Page för ett recept där jag kan displaya informationen
			// Redirecta användaren när man lyckats spara ett recept till databasen till det receptet

		}

		return (
			<div>
				<Header>
					<input
						type="text"
						value={this.state.title}
						onChange={this.saveToState}
						name="title"
						required
						placeholder="Nytt recept..."
					/>
					<i onClick={this.editTitle} className="icofont-ui-edit" />
				</Header>
				<RecipeForm>{ActiveComponent}</RecipeForm>
			</div>
		);
	}
}

export default NewRecipeForm;
