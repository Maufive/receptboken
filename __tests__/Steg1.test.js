import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Steg1 from "../components/Steg1";
import NewRecipeForm from "../components/NewRecipeForm";
import { fakeUser } from "../testUtils";

describe("<Steg 1/>", () => {
	let wrapper;
	beforeEach(() => (wrapper = shallow(<Steg1 />)));

	// renders the component and matches snapshot
	it("renders the component and matches the snapshot", () => {
		expect(wrapper.find("FormStyles__StyledForm").exists()).toBe(true);
		expect(toJSON(wrapper)).toMatchSnapshot();
	});

	// check that inputs are rendering and responding to onChange event
	describe("Title input", () => {
		const ParentComponent = mount(<NewRecipeForm user={fakeUser} />);
		it("renders without error", () => {
			expect(wrapper.find("#title").length).toEqual(1);
		});

		it("responds to the onChange event", () => {
			ParentComponent.find("#title").simulate("change", {
				target: { name: "title", value: "Test Title" }
			});
			expect(ParentComponent.state("title")).toEqual("Test Title");
		});
	});

	describe("Ingredients inputs", () => {
		it("renders without error", () => {
			expect(wrapper.find("#servings").length).toEqual(1);
			expect(wrapper.find("#item").length).toEqual(1);
			expect(wrapper.find("#numberOfUnits").length).toEqual(1);
			expect(wrapper.find("#units").length).toEqual(1);
		});

		it("responds to the onChange event", () => {
			wrapper.find("#servings").simulate("change", {
				target: { name: "servings", value: "6" }
			});

			wrapper.find("#item").simulate("change", {
				target: { name: "item", value: "Spaghetti" }
			});
			wrapper.find("#numberOfUnits").simulate("change", {
				target: { name: "numberOfUnits", value: 10 }
			});
			wrapper.find("#units").simulate("change", {
				target: { name: "units", value: "tsk" }
			});

			expect(wrapper.state("servings")).toEqual("6");
			expect(wrapper.state().item.input).toEqual("Spaghetti");
			expect(wrapper.state().item.numberOfUnits).toEqual(10);
			expect(wrapper.state().item.units).toEqual("tsk");
		});
	});

	describe("Displays a list of ingredients", () => {
		it("does not render when there are no ingredients in state", () => {
			expect(wrapper.find("List").length).toEqual(0);
		});

		it("does render a list of ingredients when there are ingredients in state", () => {
			const data = [
				{
					input: "Spaghetti",
					numberOfUnits: 1,
					units: "dl"
				}
			];

			wrapper.setState({ arr: data });
			wrapper.update();
			expect(wrapper.find("Steg1Styles__List").length).toEqual(1);
		});
	});
});
