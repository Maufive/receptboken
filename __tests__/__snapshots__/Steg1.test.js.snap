// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Steg 1/> renders the component and matches the snapshot 1`] = `
<FormStyles__StyledForm>
  <h3>
    <span>
      Ingredienser
    </span>
    <span>
      Steg 
      1
       av 3
    </span>
  </h3>
  <div>
    <input
      id="title"
      name="title"
      required={true}
      type="text"
    />
    <label
      htmlFor="title"
    >
      <EditIcon
        viewBox="0 0 55.25 55.25"
        xmlns="http://www.w3.org/2000/svg"
      />
       Titel
    </label>
  </div>
  <div>
    <p>
      Portioner:
    </p>
    <select
      id="servings"
      name="servings"
      onChange={[Function]}
    >
      <option
        value="2"
      >
        2
      </option>
      <option
        value="4"
      >
        4
      </option>
      <option
        value="6"
      >
        6
      </option>
      <option
        value="8"
      >
        8
      </option>
    </select>
  </div>
  <div>
    <input
      id="item"
      name="item"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <label
      htmlFor="item"
    >
      <IngredientIcon
        style={
          Object {
            "fill": "#15BD76",
          }
        }
        viewBox="0 0 56.823 56.823"
        xmlns="http://www.w3.org/2000/svg"
      />
       Fyll i ingredienser:
    </label>
  </div>
  <div
    style={
      Object {
        "marginBottom": "1rem",
      }
    }
  >
    <p>
      Antal: 
    </p>
    <input
      id="numberOfUnits"
      min={0}
      name="numberOfUnits"
      onChange={[Function]}
      type="number"
      value={1}
    />
  </div>
  <div
    style={
      Object {
        "marginBottom": "1rem",
      }
    }
  >
    <p>
      Enhet: 
    </p>
    <select
      id="units"
      name="units"
      onChange={[Function]}
      value="st"
    >
      <option
        defaultValue="st"
      >
        Styck
      </option>
      <option
        value="gram"
      >
        Gram
      </option>
      <option
        value="msk"
      >
        Matsked
      </option>
      <option
        value="tsk"
      >
        Tesked
      </option>
      <option
        value="dl"
      >
        Deciliter
      </option>
      <option
        value="frp"
      >
        Förpackning
      </option>
      <option
        value="ml"
      >
        Milliliter
      </option>
    </select>
  </div>
  <Button
    fullWidth={true}
    onClick={[Function]}
  >
    <AddBoldIcon
      viewBox="0 0 42 42"
      xmlns="http://www.w3.org/2000/svg"
    />
    Lägg till
  </Button>
  <Steg1Styles__List />
  <Button__DisabledButton
    disabled={true}
    fullWidth={true}
  >
    Nästa steg →
  </Button__DisabledButton>
</FormStyles__StyledForm>
`;
